packages/theme-addresses/addressForm.ts

```diff
@@ -0,0 +1,274 @@
+import {loadCountries, Country} from './loader';
+import { FromEventTarget } from 'rxjs/internal/observable/fromEvent';
+
+var COUNTRIES: Country[];
+var FIELD_REGEXP = /({\w+})/g;
+var LINE_DELIMITER = '_';
+var INPUT_SELECTORS = {
+  lastName: '[name="address[last_name]"]',
+  firstName: '[name="address[first_name]"]',
+  company: '[name="address[company]"]',
+  address1: '[name="address[address1]"]',
+  address2: '[name="address[address2]"]',
+  country: '[name="address[country]"]',
+  zone: '[name="address[province]"]',
+  postalCode: '[name="address[zip]"]',
+  city: '[name="address[city]"]',
+  phone: '[name="address[phone]"]',
+};
+
+interface FormElement {
+  wrapper?: HTMLElement;
+  input?: HTMLInputElement | HTMLSelectElement;
+  labels?: HTMLElement;
+}
+
+interface FormElements {
+  lastName?: FormElement;
+  firstName?: FormElement;
+  company?: FormElement;
+  address1?: FormElement;
+  address2?: FormElement;
+  country?: FormElement;
+  zone?: FormElement;
+  postalCode?: FormElement;
+  city?: FormElement;
+  phone?: FormElement;
+}
+
+interface AddressFormOptions {
+  loaded?(): void;
+}
+
+export default function AddressForm(
+  rootEl: HTMLElement,
+  locale: string = 'en',
+  options: AddressFormOptions = {},
+) {
+  var formElements = loadFormElements(rootEl);
+  validateElements(formElements);
+
+  if (COUNTRIES) {
+    init(rootEl, formElements, options);
+  } else {
+    loadCountries(locale).then(function(countries) {
+      COUNTRIES = countries;
+      init(rootEl, formElements, options);
+    });
+  }
+}
+
+/**
+ * Runs when countries have been loaded
+ */
+function init(rootEl: HTMLElement, formElements: FormElements, options: AddressFormOptions) {
+  populateCountries(formElements);
+  var selectedCountry = formElements.country.input
+    ? formElements.country.input.value
+    : null;
+  handleCountryChange(rootEl, formElements, selectedCountry);
+  if (options.loaded) {
+    options.loaded();
+  }
+}
+
+/**
+ * Handles when a country change: set labels, reorder fields, populate zones
+ */
+function handleCountryChange(rootEl: HTMLElement, formElements: FormElements, countryCode: string) {
+  var country = getCountry(countryCode);
+
+  setEventListeners(rootEl, formElements, country);
+  setLabels(formElements, country);
+  reorderFields(rootEl, formElements, country);
+  populateZones(formElements, country);
+}
+
+/**
+ * Sets up event listener for country change
+ */
+function setEventListeners(rootEl: HTMLElement, formElements: FormElements, country: Country) {
+  formElements.country.input.addEventListener('change', function(event) {
```

Comment:
You addEventListener whenever the country selector changes. Doesn't this result in multiple event handlers being built up and repeating the task more than needed?

Reply:
Oh true, good catch 👍